### ISSUE TITLE
scope4code: cscope is not installed (or not added to PATH)

### ISSUE THREAD

**Issue by MileanCo** [xulion/scope4code#35]
Using this extension on Windows. After updating my path (both user and system) I can run "cscope" from the Command Prompt. However, VSCode still reports: 
`scope4code: cscope is not installed (or not added to PATH)`

I even put 'cscope' both 32bit and 64bit in the C:\Windows\System32\ and this plugin cant find it anywhere.... Any ideas?

EDIT: Also having this problem https://github.com/Microsoft/vscode/issues/62953
I am on a locked down system with no internet access.

**Comment by xulion** [comment]
Pull #37 fixed windows cscope. v0.0.14 has been released with the fix

**Comment by MileanCo** [comment]
Thanks for fixing this! I switched to Linux now, but I actually still get this error... "cscope" is accessible from the terminal in bash, csh, and ksh. What shell/terminal does this use?? My cscope binary is in ~/bin/ because I cant put it in /usr/bin

**Comment by MileanCo** [comment]
Here's the ENV that the process is starting with to find cscope. I noticed my PATH here never got updated to what I have set in .cshrc and .bashrc, so obviously it cant find it.

`{"PAM_KWALLET5_LOGIN":"/tmp/kwallet5_jstenkvi.socket","USER":"jstenkvi","LANGUAGE":"en_US.UTF-8","XDG_SEAT":"seat0","XDG_SESSION_TYPE":"x11","COMPIZ_CONFIG_PROFILE":"mate","SSH_AGENT_PID":"10265","SHLVL":"0","HOME":"/u/jstenkvi","QT4_IM_MODULE":"xim","DESKTOP_SESSION":"mate","QT_STYLE_OVERRIDE":"gtk","QT_LINUX_ACCESSIBILITY_ALWAYS_ON":"1","GTK_MODULES":"gail:atk-bridge:canberra-gtk-module:topmenu-gtk-module","XDG_SEAT_PATH":"/org/freedesktop/DisplayManager/Seat0","DBUS_SESSION_BUS_ADDRESS":"unix:abstract=/tmp/dbus-ul52u6ic0Y,guid=bd4f3efe52dd07afd3dcac4f5d66d14d","QT_QPA_PLATFORMTHEME":"appmenu-qt5","MANDATORY_PATH":"/usr/share/gconf/mate.mandatory.path","LOGNAME":"jstenkvi","GTK_IM_MODULE":"ibus","DEFAULTS_PATH":"/usr/share/gconf/mate.default.path","XDG_SESSION_ID":"c2","CLUTTER_BACKEND":"x11","GTK_OVERLAY_SCROLLING":"0","GTK2_MODULES":"overlay-scrollbar","PATH":"/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin","XDG_SESSION_PATH":"/org/freedesktop/DisplayManager/Session0","XDG_RUNTIME_DIR":"/run/user/107144","DISPLAY":":0","LANG":"","XDG_CURRENT_DESKTOP":"MATE","XMODIFIERS":"@im=ibus","XDG_SESSION_DESKTOP":"mate","XAUTHORITY":"/u/jstenkvi/.Xauthority","XDG_GREETER_DATA_DIR":"/var/lib/lightdm-data/jstenkvi","SSH_AUTH_SOCK":"/run/user/107144/keyring/ssh","SHELL":"/bin/csh","QT_ACCESSIBILITY":"1","GDMSESSION":"mate","XDG_VTNR":"7","QT_IM_MODULE":"ibus","PWD":"/u/jstenkvi","CLUTTER_IM_MODULE":"xim","XDG_CONFIG_DIRS":"/etc/xdg/xdg-mate:/etc/xdg","XDG_DATA_DIRS":"/usr/share/mate:/usr/share/mate:/usr/local/share:/usr/share:/var/lib/snapd/desktop","MATE_DESKTOP_SESSION_ID":"this-is-deprecated","SESSION_MANAGER":"local/rtpc14p15u-0054:@/tmp/.ICE-unix/10051,unix/rtpc14p15u-0054:/tmp/.ICE-unix/10051","GIO_LAUNCHED_DESKTOP_FILE":"/usr/share/applications/code.desktop","GIO_LAUNCHED_DESKTOP_FILE_PID":"32305","VSCODE_IPC_HOOK":"/run/user/107144/vscode-9ad050ca-1.37.1-main.sock","VSCODE_NLS_CONFIG":"{\"locale\":\"en-us\",\"availableLanguages\":{},\"_languagePackSupport\":true}","VSCODE_LOGS":"/u/jstenkvi/.config/Code/logs/20190903T171126","AMD_ENTRYPOINT":"vs/workbench/services/extensions/node/extensionHostProcess","PIPE_LOGGING":"true","VERBOSE_LOGGING":"true","VSCODE_IPC_HOOK_EXTHOST":"/tmp/vscode-ipc-13c16940-98b4-48c8-a7f2-ad1a52fbe22c.sock","VSCODE_HANDLES_UNCAUGHT_ERRORS":"true","VSCODE_LOG_STACK":"false","VSCODE_PREVENT_FOREIGN_INSPECT":"true","ELECTRON_RUN_AS_NODE":"1","APPLICATION_INSIGHTS_NO_DIAGNOSTIC_CHANNEL":"true"}`

**Comment by xulion** [comment]
The implementation is using child_process.spawn so it shall inherit whatever vscode is using. It's wierd since you can access cscope from the terminal opened from vscode but not from it's extension env.

Does previous version 0.0.13 works for you? just in case I broke anything as I'm trying to redesign many things and still in progress.

PS. windows is pretty slow to run cscope.

**Comment by MileanCo** [comment]
Well, even if I hack the code myself and just set the path manually:
`process.env.PATH = "..." `

it shows my changed path in an error message I added, but the spawnSync still doesnt work. If I add the full absolute path to "cscope" such as:
`const ret = spawnSync("/u/jstenkvi/bin/cscope", ['-V'], cscopeExecConfig);`

the 'ret' turns back a bunch of random crap and doesnt work:

`scope4code: {"status":0,"signal":null,"output":[null,{"type":"Buffer","data":[]},{"type":"Buffer","data":[47,120,47,101,110,103,47,114,97,116,115,47,97,114,99,104,47,105,51,56,54,45,114,101,100,104,97,116,45,108,105,110,117,120,56,46,48,47,98,105,110,47,99,115,99,111,112,101,45,49,53,46,54,45,114,101,97,108,58,32,118,101,114,115,105,111,110,32,49,53,46,54,10]}],"pid":3801,"stdout":{"type":"Buffer","data":[]},"stderr":{"type":"Buffer","data":[47,120,47,101,110,103,47,114,97,116,115,47,97,114,99,104,47,105,51,56,54,45,114,101,100,104,97,116,45,108,105,110,117,120,56,46,48,47,98,105,110,47,99,115,99,111,112,101,45,49,53,46,54,45,114,101,97,108,58,32,118,101,114,115,105,111,110,32,49,53,46,54,10]}}`

Could it be that i'm on Ubuntu 64bit, and this is a 32bit cscope binary? It works on the command line just fine though on this system.

**Comment by xulion** [comment]
Joel,

The output is ascii and when translate it to characters it actually showing below:
"/x/eng/rats/arch/i386-redhat-linux8.0/bin/cscope-15.6-real: version 15.6\n"

What system are you using maybe I can setup an environment to reproduce what you are facing and figuring out a way to fix it.

**Comment by MileanCo** [comment]
I'm running on a Ubuntu host with no internet access (VM). I'm stuck using this old cscope-15.6 version because the database for the codebase Im trying to use this for is massive and has some extra steps added to build the cscope DB... As such, I've created a symlink in the `project_root_folder_/.vscode/cscope/cscope.out --> (real_path)`

I was able to get past the checkTool() method by changing this:
`if (ret.stderr.toString().search("cscope") >= 0) { ... }
`

However, now I got stuck on actually running the tool, which returns with this. Probably because I either put the cscope.out symlink in the wrong place or the wrong path is added to cscope command. I also noticed I run typically cscope with -d -f to skip updating cross-reference. Maybe adding a section in the extension options to add extra flags to cscope would be nice. 
`scope4code: cscope: no source files found
`

I can probably continue hacking at this to get a very temporary solution but if you can reproduce this and get a fix that would be fantastic :) .

**Comment by xulion** [comment]
I did not realize that the version output from a self built cscope could be this different :-(. I will update implementation for cscope check and also allow to put a path in the configure so you don't have to struggle the env PATH. (need to figure out a way to test this)

For the result, it's better keep the database not moved and update database path in the config, hopefully it could solve your issue. Also if you use F1 and type in "cscope" and search there, you might see what is actually returned from the search.

**Comment by MileanCo** [comment]
Okay thanks! You could just temporarily move cscope from out of your env path so vscode cant find it, then use this new field for the path and see if it works.

Where is this database path config you are talking about? I dont see it under the "Contributions" section of the Extension. I only saw it in the code in CscopeExecutor.js
```
class CscopeExecutor {
    constructor(source_paths, database_path, build_command, out) {
        this.source_paths = source_paths;
        //this.database_path = database_path;
        this.database_path = "/u/joel/cscope";
```

**Comment by xulion** [comment]
there is a config file under .vscode folder
.vscode/cscope_conf.json

The content shall looks like below:
{
    "version": "0.0.13",
    "open_new_column" : "no",
    "engine_configurations": [
        {
            "cscope" : {
                "paths" : [
                    "${workspaceRoot}"
                ],
                "build_command" : "",
                "database_path" : "${workspaceRoot}/.vscode/cscope"
            }
        }
    ]
}

**Comment by MileanCo** [comment]
Ok! It doesnt seem to matter where the DB is loaded, the issue is the same that the path is wrong bc parts of it are repeated. The **bold** part is repeated, because that is the folder I loaded in vscode within the workspace (the db was built from dev/).

cannot open file:///u/jstenkvi/p4/dev/**nas/prod/common**/nas/prod/common/track/qos.c. Detail: Unable to read file (Error: File not found (/u/jstenkvi/p4/dev/**nas/prod/common**/nas/prod/common/track/qos.c))

The references it finds are correct though on the right-side display (it shows only _nas/prod/common/track/qos.c_ for example).

**Comment by xulion** [comment]
can you try let the extension to build the database for you? Press F1 then type cscope you shall see the build option.

file DefinitionProvider.ts, line 34 has the code to generate the full path: 
                fileName = path.join(vscode.workspace.rootPath, fileName);

**Comment by MileanCo** [comment]
Yeah, it works if I let it build its own database! I can do "Find all references" and click on the results and it takes me there. Nice! 
But when I want to step-into a function I have to press F1->"Cscope: find this function definition" and instead of going to where it's defined, it opens a text window with the path... Maybe this is normal behaviour but it would be nice if it just went to the definition as expected (in vim+cscope you can just press CTRL+] and it goes directly to the definition, same in the standard C/C++ VSCode Extension).

**Comment by xulion** [comment]
Just released new version. Now in setting (vscode file->preferences->settings) you can set path of executables (~/ is not allowed somehow, it has to be full path). Also I updated cscope -V result parsing. It shall now work with your scope. I'm closing this issue now.

Regarding to the direct definition navigation, I would create an enhancement issue for it. The reason it still show a list is because the code base I'm dealing with would have multiple implementations for one function. Some due to the mock version for test and some due to the overridden C++ function (I'm using it for a C++ project). Need to figure out a good way to support this.

**Comment by MileanCo** [comment]
Where did you add this setting? Ive tried the new update and added the new executablePath in vscode settings.json but it still doesnt work:
"scope4code.executablePath": "/u/milean/bin/cscope"


<img width="1001" alt="Screen Shot 2019-09-20 at 12 05 32 PM" src="https://user-images.githubusercontent.com/6936016/65341723-fa2bc900-db9e-11e9-94eb-49935c32b10b.png">

**Comment by xulion** [comment]
Can you try use workspace setting?

**Comment by MileanCo** [comment]
It doesnt work there either unfortunately :( 
Here is my settings.json:
{
    "update.mode": "none",
    "telemetry.enableCrashReporter": false,
    "telemetry.enableTelemetry": false,
    "update.showReleaseNotes": false,
    "extensions.autoUpdate": false,
    "terminal.external.linuxExec": "/bin/bash",
    "terminal.explorerKind": "external",
    "terminal.integrated.shell.linux": "/bin/bash",
    "zenMode.hideTabs": false,
    "editor.tabSize": 8,
    "scope4code.executablePath": "/u/milean/bin/cscope"
}

My workspace config looks kinda like this:
{
    "folders": [
        { "name": "milean/prod/common", "path": "./milean/prod/common/" }
    ],
    "settings": {
        "perforce.activationMode": "always",
        "perforce.dir": "/x/eng/bbrtp/users/milean/dev",
        "perforce.client": "milean:dev",
        "terminal.integrated.cwd": "/x/eng/bbrtp/users/milean/dev",
        "scope4code.executablePath": "/u/milean/bin/cscope"
    }
}

/u/milean/bin/cscope is a sym-link to the actual location, but adding the real location doesnt work either.

**Comment by xulion** [comment]
I removed my cscope path from PATH and with scope4code.executablePath added to <workspace>/.vscode/settings.json, everything just works fine.

Is settings.json you showed above user setting or workspace setting?

**Comment by MileanCo** [comment]
The user setting is the first one (~/.config/Code/User/settings.json), workspace setting ( is 2nd one.  I'm on Ubuntu. Is there a way I can open the console output and see any errors?

**Comment by xulion** [comment]
curious how your work space setting looks like that. Is it generated by vscode? My workspace setting is \<workspace\>/.vscode/settings.json and scope4code.executablePath is not under setting. Looks something like this:
```
{
    //other settings
    "scope4code.executablePath": "/u/milean/bin/cscope"
}
```

**Comment by xulion** [comment]
Oh, btw, is /u/milean/bin/cscope the binary path name or it's the folder name where cscope exe is?

**Comment by MileanCo** [comment]
It's the path straight to the executable

**Comment by MileanCo** [comment]
It worked now! Maybe the instructions should say that it's the path to the folder where cscope binary is located, not the "executable path". Using it is way slower now than before though, I think I added or removed a flag to cscope to make it faster, so I may have to edit the code to do that again.

**Comment by xulion** [comment]
Cool. Will update the description. I am so use to C and when say path it means folder. LOL.

It's a bit slow for the first run after you make changes in the code. the parameter now will update the database. Later run shall be fast.

I'm working on engine config now (actually you shall be able to see the setting "Engine Config" already, just it won't take effect now). It looks like below and once it's done, you shall be able to do whatever you want, even you can use engine other than cscope

```
            "scope4code.engineConfig": {
                "type": "object",
                "default": {
                    "platform_index" : {
                        "win32" : 1,
                        "linux" : 0,
                        "mac_os" : 0,
                        "others" : 0
                    },
                    "config" : [
                        {
                            "find_cmd" : "find ${src_path} -type f -name *.c -o -type f -name *.h -o -type f -name *.cpp -o -type f -name *.cc -o -type f -name *.mm",
                            "database_cmd" : "cscope -b -q -k",
                            "find_all_ref" : "cscope -q -L0 ${text}",
                            "find_define" : "cscope -q -L1 ${text}",
                            "find_callee" : "cscope -q -L2 ${text}",
                            "find_caller" : "cscope -q -L3 ${text}",
                            "find_text" : "cscope -q -L4 ${text}"
                        },
                        {
                            "find_cmd" : "cmd /C dir /s/a/b ${src_path}\\*.c ${src_path}\\*.h ${src_path}\\*.cpp ${src_path}\\*.cc ${src_path}\\*.mm",
                            "database_cmd" : "cscope -b -q -k",
                            "find_all_ref" : "cscope -q -L0 ${text}",
                            "find_define" : "cscope -q -L1 ${text}",
                            "find_callee" : "cscope -q -L2 ${text}",
                            "find_caller" : "cscope -q -L3 ${text}",
                            "find_text" : "cscope -q -L4 ${text}"
                        }
                    ]
                }
            }
```

**Comment by MileanCo** [comment]
> Cool. Will update the description. I am so use to C and when say path it means folder. LOL.
> 
> It's a bit slow for the first run after you make changes in the code. the parameter now will update the database. Later run shall be fast.
> 
> I'm working on engine config now (actually you shall be able to see the setting "Engine Config" already, just it won't take effect now). It looks like below and once it's done, you shall be able to do whatever you want, even you can use engine other than cscope
> 
> ```
>             "scope4code.engineConfig": {
>                 "type": "object",
>                 "default": {
>                     "platform_index" : {
>                         "win32" : 1,
>                         "linux" : 0,
>                         "mac_os" : 0,
>                         "others" : 0
>                     },
>                     "config" : [
>                         {
>                             "find_cmd" : "find ${src_path} -type f -name *.c -o -type f -name *.h -o -type f -name *.cpp -o -type f -name *.cc -o -type f -name *.mm",
>                             "database_cmd" : "cscope -b -q -k",
>                             "find_all_ref" : "cscope -q -L0 ${text}",
>                             "find_define" : "cscope -q -L1 ${text}",
>                             "find_callee" : "cscope -q -L2 ${text}",
>                             "find_caller" : "cscope -q -L3 ${text}",
>                             "find_text" : "cscope -q -L4 ${text}"
>                         },
>                         {
>                             "find_cmd" : "cmd /C dir /s/a/b ${src_path}\\*.c ${src_path}\\*.h ${src_path}\\*.cpp ${src_path}\\*.cc ${src_path}\\*.mm",
>                             "database_cmd" : "cscope -b -q -k",
>                             "find_all_ref" : "cscope -q -L0 ${text}",
>                             "find_define" : "cscope -q -L1 ${text}",
>                             "find_callee" : "cscope -q -L2 ${text}",
>                             "find_caller" : "cscope -q -L3 ${text}",
>                             "find_text" : "cscope -q -L4 ${text}"
>                         }
>                     ]
>                 }
>             }
> ```

Nice! When do you think this will be available? I'd like to try this plugin again, but only if I can modify the cscope flags to make it usable (faster) for me.

**Comment by xulion** [comment]
Just published version 0.1.2. Now all command are customizable!

**Comment by xulion** [comment]
Joel,
Glad you figured that out. Yes now you need reload the workspace as the config loads with extension starts and it does not reload after it's changed. I will improve this later as now I have limited time so I can only add one feature at a time. 

Next step would be reload configure. After that I will try to find a better way integrate the result to vscode UI. I'm currently thinking of something like all other language provider allow you to use ctrl+click to find the definition. Also because of the filelist I shall be able to improve the ctrl+E (find file) for large project.

Will try to make all above happen hopefully this year.
